rse-r-package.Rmd

```diff
@@ -0,0 +1,590 @@
+
+### Exercise: Exploring a package
+
+1.  Clone a GitHub repository that contains an R package, such as <https://github.com/r-lib/usethis>.
+2.  Which files in this repository do you recognize, and what are they for?
+3.  Which files are there to satisfy something other than R's packaging system?
+
+## How do I document a R package? {#rse-r-package-document}
+
+Our next task is to document our function.
+To do this,
+we turn to Hadley Wickham's *[R Packages][wickham-packages]*
+and Karl Broman's "[R package primer][broman-packages]"
+for advice on writing roxygen2.
+We then return to our source file and put a specially-formatted comment in front of our code:
+
+```{r, child="rse-r-package/etc/frequency-doc.R"}
+```
+
+This comment is an example of [embedded documentation](glossary.html#embedded-documentation).
+Over the years,
+programmers have found that if they put code in one file and documentation in another,
+the documentation quickly falls out of date with the code:
+people will change a function's name or add a new parameter and forget to update the docs.
+If the documentation lives right beside the code,
+on the other hand,
+the next person to modify the code is far more likely to remember to update it.
+Tools like roxygen2 can read the code,
+extract the documentation,
+and format it as HTML or PDF.
+They can also do things like create an index,
+which would be even more painful to do by hand than writing the documentation itself.
+
+roxygen2 processes comment lines that start with `#'` (hash followed by single quote).
+Putting a comment block right before a function associates that documentation with that function,
+and `@something` indicates a roxygen2 command,
+so what this file is saying is:
+
+-   the function has two parameters called `input_file` and `output_file`;
+-   it returns the number of distinct words found; and
+-   we want to export it (i.e., we want it to be visible outside the package).
+
+Our function is now documented,
+but when we run `Check`,
+we still get a warning.
+After a bit more searching and experimentation,
+we discover that we need to load the `devtools` package in the console
+and run `devtools::document()` to regenerate documentation:
+it isn't done automatically.
+
+```{r eval=FALSE}
+library(devtools)
+devtools::document()
+```
+
+When we do this,
+we get good news and bad news:
+
+```
+Updating zipffreq documentation
+First time using roxygen2. Upgrading automatically...
+Updating roxygen version in /Users/gvwilson/merely/rse-r-package/zipffreq/DESCRIPTION
+Loading zipffreq
+Warning: The existing 'NAMESPACE' file was not generated by roxygen2, and will not be overwritten.
+Writing count_words.Rd
+```
+
+Ah---the tutorials did warn us about this.
+We need to delete `NAMESPACE` and re-run `devtools:document()` again:
+
+```
+Updating zipffreq documentation
+Writing NAMESPACE
+Loading zipffreq
+Writing NAMESPACE
+```
+
+Doing this creates this file:
+
+```{r, child="rse-r-package/zipffreq/NAMESPACE"}
```

Comment:
Is there any particular reason to use child documents rather than simply inserting the text here? It can be a bit disorientating having to go up and down when reviewing.

Reply:
Ensures that everything in the `.Rmd` is up to date with what's on disk (I've been bitten so... many... times...)

Reply:
Yes, I understand now given that you are keeping the R package inside the repo. Closing this.